packages/block-library/src/table-of-contents/edit.js

```diff
@@ -15,127 +15,228 @@ import {
-			// Unfortunately, this implementation only accounts for Page Break and
-			// Classic blocks, so if there are any <!--nextpage--> tags in any
-			// other block, they won't be counted. This will result in the table
-			// of contents showing headings from the wrong page if
-			// onlyIncludeCurrentPage === true. Thankfully, this issue only
-			// affects the editor implementation.
-			let page = 1;
-			for ( let i = 0; i < blockIndex; i++ ) {
-				const blockName = getBlockName( blockOrder[ i ] );
+			// Disable reason: blocks can be loaded into a *non-post* block
+			// editor, so to avoid declaring @wordpress/editor as a dependency, we
+			// must access its store by string. When the store is not available,
+			// editorSelectors will be null, and the block's saved markup will
+			// lack permalinks.
+			// eslint-disable-next-line @wordpress/data-no-store-string-literals
+			const editorSelectors = select( 'core/editor' );
+
+			const pageBreakClientIds = getGlobalBlocksByName( 'core/nextpage' );
+
+			const isPaginated = pageBreakClientIds.length !== 0;
+
+			/**
+			 * Get the relative indices of the block from top to bottom nesting
+			 * level.
+			 *
+			 * @param {string} blockClientId
+			 *
+			 * @return {number[]} The path of indices to the block.
+			 */
+			function getBlockPath( blockClientId ) {
+				const indices = getBlockParents(
+					blockClientId
+				).map( ( ancestorId ) => getBlockIndex( ancestorId ) );
+				indices.push( getBlockIndex( blockClientId ) );
+				return indices;
+			}
+
+			// We can't use just getBlockIndex because it only returns the index relative to sibling blocks, so we have to get all the indices from top to bottom.
+			const tocPath = getBlockPath( clientId );
+
+			// Calculate the page (of a paginated post) this block is part of.
+			// Note that pageBreakClientIds may not be in the order they appear on
+			// the page, so we have to iterate over all of them.
+			let tocPage = 1;
+			for ( const pageBreakClientId of pageBreakClientIds ) {
+				if (
+					comparePathAToB(
+						tocPath,
+						getBlockPath( pageBreakClientId )
+					) < 0
+				) {
+					tocPage++;
+				}
+			}
+
+			// Get the top-level block client ids, and add them and the client ids
+			// of their children to an ordered list. We don't use
+			// getClientIdsWithDescendants since it returns ids in the wrong order.
+			const allBlockClientIds = [];
+			for ( const blockClientId of getBlockOrder() ) {
+				allBlockClientIds.push(
+					blockClientId,
+					...getClientIdsOfDescendants( [ blockClientId ] )
+				);
+			}
+
+			const _latestHeadings = [];
+
+			/** The page (of a paginated post) a heading will be part of. */
+			let headingPage = 1;
+
+			/**
+			 * If the core/editor store is available, we can add permalinks to the
+			 * generated table of contents. This variable will be a link to
+			 * current post including pagination query if necessary.
+			 */
+			const permalink = editorSelectors?.getPermalink() ?? null;
```

Comment:
Wouldn't this cache the permalink available when the post is saved? If the URL structure changes later on, the ToC would still be rendered using the old one permalinks, causing a full page reload on click.

Reply:
The changed-permalinks problem also applies to links in Paragraphs, Headings, and pretty much every other static block. That said, while the user may expect their own links to break when they change permalinks on their site, they may have different expectations for the Table of Contents block.

When I started this particular PR, I was of the mindset that static rendering would be best, since it's more performant, and I could share a lot of the implementation between editor and front-end. The dynamically-rendered version currently in `trunk` was mostly of my own design, and it was a pretty convoluted implementation in hindsight. As I was attempting to handle all headings and page break tags regardless of origin, the block was building the entire post in memory and parsing it for data on every page load.

But simply using dynamic rendering for the purpose of converting already-saved heading data to markup and getting the post's current permalink sounds far more reasonable.

Does anyone have any objections or dittos for switching back to dynamic rendering?